allennlp/nn/decoding/decoder_state.py

```diff
@@ -0,0 +1,69 @@
+from typing import Generic, List, TypeVar
+
+import torch
+
+# Note that the bound here is `DecoderState` itself.  This is what lets us have methods that take
+# lists of a `DecoderState` subclass and output structures with the subclass.  Really ugly that we
+# have to do this generic typing _for our own class_, but it makes mypy happy and gives us good
+# type checking in a few important methods.
+T = TypeVar('T', bound='DecoderState')
+
+class DecoderState(Generic[T]):
+    """
+    Represents the (batched) state of a transition-based decoder.
+
+    There are two different kinds of batching we need to distinguish here.  First, there's the
+    batch of training instances passed to ``model.forward()``.  We'll use "batch" and
+    ``batch_size`` to refer to this through the docs and code.  We additionally batch together
+    computation for several states at the same time, where each state could be from the same
+    training instance in the original batch, or different instances.  We use "group" and
+    ``group_size`` in the docs and code to refer to this kind of batching, to distinguish it from
+    the batch of training instances.
+
+    So, using this terminology, a single ``DecoderState`` object represents a `grouped` collection
+    of states.  Because different states in this group might finish at different timesteps, we have
+    methods and member variables to handle some bookkeeping around this, to split and regroup
+    things.
+
+    Parameters
+    ----------
+    batch_indices : ``List[int]``
+        Our internal variables (like scores, action histories, hidden states, whatever) are
+        `grouped`, and our ``group_size`` is likely different from the original ``batch_size``.
+        This variable keeps track of which batch instance each group element came from (e.g., to
+        know what the correct action sequences are, or which encoder outputs to use).
+    action_history : ``List[List[int]]``
+        The list of actions taken so far in this state.  This is also grouped, so each state in the
+        group has a list of actions.
+
+        The type annotation says this is an ``int``, but none of the training logic relies on this
+        being an ``int``.  In some cases, items from this list will get passed as inputs to
+        ``DecodeStep``, so this must return items that are compatible with inputs to your
+        ``DecodeStep`` class.
+    score : ``List[torch.autograd.Variable]``
+        This state's score.  It's a variable, because typically we'll be computing a loss based on
+        this score, and using it for backprop during training.  Like the other variables here, it's
```

Comment:
`s/"Like the other variables here, it's grouped"/"The list length corresponds to the number of groups"/`

Reply:
Done.